extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,167 @@
+import {isSecureUrl, parseUrl} from '../../../src/url';
+
+/** Timeout in millis */
+const DEFAULT_RTC_TIMEOUT = 1000;
+
+/** @type {string} */
+const TAG = 'real-time-config';
+
+/** @type {number} */
+const MAX_RTC_CALLOUTS = 5;
+
+export class RealTimeConfigManager {
+  constructor(element, win, ampDoc) {
+    this.element = element;
+    this.win = win;
+    this.rtcConfig = null;
+    this.calloutUrls = [];
+    this.urlReplacements_ = Services.urlReplacementsForDoc(ampDoc);
+  }
+
+  executeRealTimeConfig(customMacros) {
+    if (!this.rtcConfig) {
+      return null;
+    }
+    const rtcTimeout = this.rtcConfig.timeoutMillis || DEFAULT_RTC_TIMEOUT;
+    this.inflatePublisherUrls(customMacros);
+    this.inflateVendorUrls();
+    if (this.calloutUrls.length == 0) {
+      return Promise.resolve();
+    }
+    let rtcTime;
+    return Promise.all(this.calloutUrls.map(url => {
+      const rtcStartTime = Date.now();
+      /** ISSUE: For very short timeouts, and/or for slow devices, the calculated time
+       * that RTC took (rtcTime below) can be longer than the timeout without getting timed-out,
+       * due to the time it takes to set up the timeout promise.
+       * EXAMPLE, let rtcTimeout = 100ms.
+       * Time = 0: const rtcStartTime = 0
+       * Time = 10: call Services.timerFor.timeoutPromise
+       * Time = 25: timeout promise starts the xhr callout, timeout promise sees start time as 25
+       * Time = 120: xhr callout returns, timeoutPromise sees total time as 95ms, so no timeout
+       * Time = 125: rtcTime = 125ms, which is greater than the publisher specified timeout.
+       */
+      return Services.timerFor(this.win).timeoutPromise(
+          rtcTimeout,
+          Services.xhrFor(this.win).fetchJson(
+              url, {credentials: 'include'}).then(res => {
+                // Non-200 status codes are forbidden for RTC.
+                // TODO: Add to fetchResponse the ability to
+                // check for redirects as well.
+                if (res.status != 200) {
+                  rtcTime = Date.now() - rtcStartTime;
+                  return {rtcTime, error: 'Non-200 Status', hostname};
+                }
+                return res.text().then(text => {
+                  rtcTime = Date.now() - rtcStartTime;
+                  const hostname = parseUrl(url).hostname;
+                  // An empty text response is fine, just means
+                  // we have nothing to merge.
+                  if (!text) {
+                    return {rtcTime, hostname};
+                  }
+                  const rtcResponse = tryParseJson(text);
+                  return rtcResponse ? {rtcResponse, rtcTime, hostname} :
+                  {rtcTime, hostname, error: 'Unparsable JSON'};
+                });
+              })).catch(err => {
+                const hostname = parseUrl(url).hostname;
+                const rtcTime = Date.now() - rtcStartTime;
+                return {error: err, rtcTime, hostname};
```

Comment:
change parameter name to error.  I would just make this:
catch(error => { return {
   error, rtcTime: Date.now() - rtcStartTime, hostname: parseUrl(url).hostname
}});